torchopt/distributed/autograd.py

```diff
@@ -0,0 +1,129 @@
+# limitations under the License.
+# ==============================================================================
+"""Distributed Autograd."""
+
+from threading import Lock
+from typing import Optional, Sequence, Tuple, Union
+
+import torch
+import torch.distributed.autograd as autograd
+from torch.distributed.autograd import context
+
+
+__all__ = ['is_available', 'context']
+
+
+LOCK = Lock()
+
+
+def is_available():
+    """Check if distributed is available."""
+    return autograd.is_available()
+
+
+if is_available():
+    # pylint: disable-next=unused-import
+    from torch.distributed.autograd import DistAutogradContext, get_gradients
+
+    def backward(
+        autograd_ctx_id: int,
+        tensors: Union[torch.Tensor, Sequence[torch.Tensor]],
+        retain_graph: bool = False,
+        inputs: Optional[Union[torch.Tensor, Sequence[torch.Tensor]]] = None,
+    ) -> None:
+        """Perform distributed backward pass for local parameters.
+
+        Computes the sum of gradients of given tensors with respect to graph leaves.
+
+        Args:
+            autograd_ctx_id: The autograd context id.
+            tensors (Sequence[Tensor] or Tensor): Tensors of which the derivative will be computed.
+            retain_graph (bool, optional): If :data:`False`, the graph used to compute the grad will
+                be freed. Note that in nearly all cases setting this option to :data:`True` is not
+                needed and often can be worked around in a much more efficient way.
+            inputs (Sequence[Tensor] or Tensor, optional): Inputs w.r.t. which the gradient be will
+                accumulated into ``.grad``. All other Tensors will be ignored. If not provided, the
+                gradient is accumulated into all the leaf Tensors that were used to compute the
+                attr::tensors.
+        """
+        if inputs is not None:
+            if isinstance(inputs, torch.Tensor):
+                inputs = (inputs,)
+            elif len(inputs) == 0:
+                raise RuntimeError("'inputs' argument to backward() cannot be empty.")
+            else:
+                inputs = tuple(inputs)
+            if not all(map(lambda t: t.requires_grad, inputs)):
+                raise RuntimeError('One of the differentiated Tensors does not require grad')
+
+        roots = [tensors] if isinstance(tensors, torch.Tensor) else list(tensors)
+        autograd.backward(autograd_ctx_id, roots=roots, retain_graph=retain_graph)
+
+        all_local_grads = autograd.get_gradients(autograd_ctx_id)
+        if inputs is not None:
+            inputs = set(inputs)  # type: ignore[assignment]
+            all_local_grads = {p: g for p, g in all_local_grads.items() if p in inputs}
+
+        with LOCK:
+            for p, g in all_local_grads.items():
+                if retain_graph and getattr(p, 'grad', None) is not None:
+                    p.grad = p.grad.add(g)
+                else:
+                    p.grad = g
```

Comment:
@JieRen98 Does this correct for `retain_graph` and `accumulate_grad=True` in [`torch.autograd.backward`](https://pytorch.org/docs/stable/_modules/torch/autograd.html#backward)? Or maybe just `p.grad = g`.